### ISSUE TITLE
(GOG Version) Crash on startup when trying to read process memory

### ISSUE THREAD

**Issue by swlody** [sigma144/witness-randomizer#3]
The application is crashing immediately on launch with the following stack trace:

```
WitnessRandomizer.exe!Memory::ThrowError() Line 107
	at C:\witness-randomizer\Source\Memory.cpp(107)
WitnessRandomizer.exe!Memory::ReadData<int>(const std::vector<int,std::allocator<int>> & offsets, unsigned __int64 numItems) Line 99
	at C:\witness-randomizer\Source\Memory.h(99)
WitnessRandomizer.exe!Memory::ReadPanelData<int>(int panel, int offset) Line 74
	at C:\witness-randomizer\Source\Memory.h(74)
WitnessRandomizer.exe!Special::ReadPanelData<int>(int panel, int offset) Line 95
	at C:\witness-randomizer\Source\Special.h(95)
WitnessRandomizer.exe!wWinMain(HINSTANCE__ * hInstance, HINSTANCE__ * hPrevInstance, wchar_t * lpCmdLine, int nCmdShow) Line 390
	at C:\witness-randomizer\App\Main.cpp(390)
```

The culprit seems to be a failure to the call to `ReadProcessMemory()` in `Memory::ReadData()`, which is failing with code `ERROR_NOACCESS`. This is regardless of whether or not the program is run with administrator permissions.

I set `lastSeed` in `WinMain` to a constant 0 so that the method in question wouldn't run, but the `ReadProcessMemory` simply crashes later with the same error code. I guess this is to be expected since it is trying to call `ReadProcessMemory` on the same offset (0x0064).

The result of `ComputeOffsets()` that is passed to `ReadProcessMemory()` doesn't look correct to me, `0x2e25_202c_6633_2f41` in one case. That number is produced because `ComputeOffsets` reads the memory 800 bytes past the current `cumulativeAddress` on the 3rd iteration through the loop, where 800 is `0x64 * 8` because`0x64` is passed as the "panel" to `Special::ReadPanelData()` from WinMain. The result of that read is a very large number for whatever reason.

**Comment by swlody** [comment]
A tangential issue is that when exceptions are thrown, there is no error message indicated to the user and the application just closes silently. This was pretty confusing for me for 5-10 minutes because I thought it was being blocked by antivirus or something - until I spun up a Visual Studio instance to see if something else was wrong.

**Comment by sigma144** [comment]
Thanks for reporting this issue. I'll definitely try and add that pop-up as it sounds like a great suggestion. My guess is what's happening is that your version of the Witness might have a different base address for the puzzles. That would explain ComputeOffsets() grabbing the wrong data. I'll look into it and try to have a possible fix done in the next week or so.

**Comment by swlody** [comment]
Thanks! I am using the GOG version. I would assume the executable would be the same as the Steam version but I guess it makes sense that something tricky is going on. Even the Steam DRM .dlls are loaded in the GOG version. I tried all the builds that are available on the GOG store but they all crashed with the same errors.

**Comment by swlody** [comment]
Thanks for the error message update! With most recent commit I get:

![image](https://user-images.githubusercontent.com/170795/73500497-d9ffe800-4390-11ea-869d-60758ecf57a3.png)

Hopefully this helps a little

**Comment by sigma144** [comment]
Hi swlody! Sorry I haven't got back to you in a while. I was finally able to get a hold of the GOG version of the game, and I verified the issue you were having. The problem seems to be that the variable for the puzzle array is in a different location on the GOG version for whatever reason. I am going to try and write a program to locate the variable in memory. I'll try to have it done this week if time allows.

**Comment by lennygold** [comment]
is this still being worked on for GOG versions?

**Comment by sigma144** [comment]
Hi guys, sorry for not responding in a while. I tried a couple weeks back to figure out what's going on in the GOG version, but without much success. I would reach out to the original mod creator, but I can't figure out how to contact him either. So I've put the GOG version on hold for now while I try to think of new solutions. I'm not ready to give up on it entirely though. Sorry for any GOG players who were excited to try it out right now. Hopefully I can make it happen sometime soon.

**Comment by sigma144** [comment]
Hi everyone, good news! I finally had a breakthrough with the GOG version. I'm going to run through the entire mod multiple times to make sure everything's working, but you guys can probably expect a working build for GOG sometime in the next few days.

**Comment by sigma144** [comment]
The GOG-compatible release (1.0.8) is ready! If you are a GOG player, please let me know if the mod works for you. I'll wait for confirmation from someone before closing this issue.

**Comment by swlody** [comment]
Works great for me, thanks so much for the fix!

**Comment by sigma144** [comment]
Excellent. Thank you for your patience!

**Comment by lennygold** [comment]
still gives me the same 2 errors, process not found in RAM when i have it closed; and when i have it open it gives me= error reading 0x11c in panel 0x64, please close the Witness and try again

**Comment by sigma144** [comment]
@lennygold Hi lennygold, it's the guy who made that Witness mod! Sorry for not responding to your comment for ages! Somehow I missed the notification that someone replied, and assumed all was well. If you get a chance, I uploaded a new release (1.1.0) that includes a script that might fix the issue you were having. Please give it a try and let me know if it worked for you.
